src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,351 @@
+    size_t num_elements_in  = 1;
+    size_t num_elements_out = 1;
+    for(int i = 0; i < DIM; ++i) {
+      n[i]          = /*(type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] :*/ fft_size_out[i];
+      inembed[i]    = fft_size_in[i];
+      onembed[i]    = fft_size_out[i];
+      num_elements_in  *= fft_size_in[i];
+      num_elements_out *= fft_size_out[i];
+    }
+
+    // Copy input to temporary buffer to perform FFTs one by one
+    DeferredBuffer<INPUT_TYPE, DIM> input_buffer(Rect<DIM>(zero, fft_size_in - one),
+                                                 Memory::GPU_FB_MEM,
+                                                 nullptr /*initial*/,
+                                                 128 /*alignment*/);
+    CHECK_CUDA(cudaMemcpyAsync(input_buffer.ptr(zero), in.ptr(zero), num_elements_in*sizeof(INPUT_TYPE), cudaMemcpyDeviceToDevice, stream));
+
+    for(auto ax = axes.begin(); ax != axes.end(); ++ax) {
+      // Create the plan
+      cufftHandle plan;
+      CHECK_CUFFT(cufftCreate(&plan));
+      CHECK_CUFFT(cufftSetAutoAllocation(plan, 0 /*we'll do the allocation*/));
+      CHECK_CUFFT(cufftSetStream(plan, stream));
+
+      // Single axis dimensions / stridfes
+      int size_1d = n[*ax];
+      // TODO: batches only correct for DIM <= 3. Fix for N-DIM case
+      int batches = (DIM == 3 && *ax == 1) ? n[2] : num_elements_in / n[*ax];
+      int istride = 1;
+      int ostride = 1;
+      for(int i = *ax+1; i < DIM; ++i) {
+        istride *= fft_size_in[i];
+        ostride *= fft_size_out[i];
+      }
+      int idist = (*ax == DIM-1) ? fft_size_in[*ax] : 1;
+      int odist = (*ax == DIM-1) ? fft_size_out[*ax] : 1;
+
+      // Create the plan and allocate a temporary buffer for it if it needs one
+      CHECK_CUFFT(cufftMakePlanMany(plan, 1, &size_1d, inembed, istride, idist, onembed, ostride, odist, (cufftType)type, batches, &workarea_size));
+
+      DeferredBuffer<uint8_t, 1> workarea_buffer;
+      if(workarea_size > 0) {
+        const Point<1> zero1d(0);
+        workarea_buffer =
+          DeferredBuffer<uint8_t, 1>(Rect<1>(zero1d, Point<1>(workarea_size - 1)),
+                                     Memory::GPU_FB_MEM,
+                                     nullptr /*initial*/,
+                                     128 /*alignment*/);
+        void* workarea = workarea_buffer.ptr(zero1d);
+        CHECK_CUFFT(cufftSetWorkArea(plan, workarea));
+      }
+
+      // TODO: following function only correct for DIM <= 3. Fix for N-DIM case
+      cufft_axes_plan<DIM, INPUT_TYPE, INPUT_TYPE>::execute(plan, input_buffer, input_buffer, out_rect, in_rect, *ax, direction);
+
+      // Clean up our resources, DeferredBuffers are cleaned up by Legion
+      CHECK_CUFFT(cufftDestroy(plan));
+    }
+    CHECK_CUDA(cudaMemcpyAsync(out.ptr(zero), input_buffer.ptr(zero), num_elements_out*sizeof(OUTPUT_TYPE), cudaMemcpyDeviceToDevice, stream));
+    CHECK_CUDA(cudaStreamDestroy(stream));
```

Comment:
Do we need a `cudaStreamSynchronize` before destroying the stream, or will `cudaStreamDestroy` wait until the stream is free of work before destroying it?

Reply:
IIRC cudaStreamDestroy waits until the work is completed before destroying, but it does not synchronize with host, so it might be necessary regardless; I don't think cuNumeric is aware of copies inside deferred buffers.